spring-integration-jms/src/main/java/org/springframework/integration/jms/JmsOutboundGateway.java

```diff
@@ -726,34 +742,44 @@ protected Object handleRequestMessage(final Message<?> message) {
 					return null;
 				}
 			}
-			Object result = jmsReply;
-			if (this.extractReplyPayload) {
-				result = this.messageConverter.fromMessage(jmsReply);
-				if (logger.isDebugEnabled()) {
-					logger.debug("converted JMS Message [" + jmsReply + "] to integration Message payload [" + result + "]");
-				}
-			}
-			Map<String, Object> jmsReplyHeaders = this.headerMapper.toHeaders(jmsReply);
 
-			if (this.replyContainer != null && this.correlationKey != null) {
-				// do not propagate back the gateway's internal correlation id
-				jmsReplyHeaders.remove(this.correlationKey);
-			}
-			Message<?> replyMessage = null;
-			if (result instanceof Message){
-				replyMessage = this.getMessageBuilderFactory().fromMessage((Message<?>) result).copyHeaders(jmsReplyHeaders).build();
+			if (reply instanceof javax.jms.Message) {
+				return replyToReplyMessage((javax.jms.Message) reply);
 			}
 			else {
-				replyMessage = this.getMessageBuilderFactory().withPayload(result).copyHeaders(jmsReplyHeaders).build();
+				return reply;
 			}
-			return replyMessage;
 		}
 		catch (JMSException e) {
 			throw new MessageHandlingException(requestMessage, e);
 		}
 	}
 
-	private javax.jms.Message sendAndReceiveWithContainer(Message<?> requestMessage) throws JMSException {
+	private Message<?> replyToReplyMessage(javax.jms.Message jmsReply) throws JMSException {
+		Object result = jmsReply;
+		if (this.extractReplyPayload) {
+			result = this.messageConverter.fromMessage(jmsReply);
+			if (logger.isDebugEnabled()) {
+				logger.debug("converted JMS Message [" + jmsReply + "] to integration Message payload [" + result + "]");
+			}
+		}
+		Map<String, Object> jmsReplyHeaders = this.headerMapper.toHeaders(jmsReply);
+
+		if (this.replyContainer != null && this.correlationKey != null) {
+			// do not propagate back the gateway's internal correlation id
+			jmsReplyHeaders.remove(this.correlationKey);
+		}
+		Message<?> replyMessage = null;
+		if (result instanceof Message){
+			replyMessage = this.getMessageBuilderFactory().fromMessage((Message<?>) result).copyHeaders(jmsReplyHeaders).build();
+		}
+		else {
+			replyMessage = this.getMessageBuilderFactory().withPayload(result).copyHeaders(jmsReplyHeaders).build();
```

Comment:
Would be better to return to the `AbstractMessageProducingHandler` just `MessageBuilder`.
The last one does exactly this:
```
if (reply instanceof Message) {
	builder = this.getMessageBuilderFactory().fromMessage((Message<?>) reply);
}
else if (reply instanceof AbstractIntegrationMessageBuilder) {
	builder = (AbstractIntegrationMessageBuilder<?>) reply;
}
```
So, it might be a bit sad to recreate message all the time.

From other side I see that I use `reply` for the `getOutputChannelFromRoutingSlip()`, so `MessageBuilder` might expose access to its `payload` and `header` on the matter...

But that is a different story and should be addressed in the separate JIRA.
Right now I just want to change this code to return Builder not the final message.

WDYT?